Demo datasource page

tinahollygb opened the pull request:
### Features and Changes

- User can create a demo datasource project
- User can see if they have a demo datasource project already
- User can delete their demo datasource project
- Updated some onboarding screens to link to it (let me know if I missed any)
- Changed backend to support experiment-ref instead of experiment for demo feature(s)


### Dependencies

- [x] https://github.com/growthbook/growthbook/pull/1487 - Once this is merged, the diff will shrink as these changes are merged into my changes


### Testing

Test creating and deleting a project.


## Screenshots

See screenshots below as well as the ones in all of the comments in the PR, which include updates based on feedback.

### In-app

<img width="1660" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/c4706a4d-f2de-43e6-9778-90498afacccd">

<img width="1660" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/e8b1b7b5-ff4c-4b34-be6c-0a438c580022">

#### Getting Started

The Getting Started page has the same UI, but when clicking, instead of hitting the old sample data endpoint, it navigates to the new demo datasource page.

![CleanShot 2023-08-11 at 11 41 29@2x](https://github.com/growthbook/growthbook/assets/113377031/270eeb7a-42c0-4ed6-8b33-464acaef0793)

#### Experiments Getting Started

The empty state for the experiments page and on the `/dashboard` page when there's no experiments.

<img width="1332" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/41fd0833-0fe1-4643-8894-90abe348f72e">

#### DataSources page

If the project isn't the demo project, and the demo datasource project doesn't already exist, they'll see a link to the demo datasource page so they can create it.

![CleanShot 2023-08-11 at 13 55 42@2x](https://github.com/growthbook/growthbook/assets/113377031/1c1964e5-b733-4d86-810c-263f186462a6)


#### Add/Edit Metric modal

<img width="866" alt="Screen Shot 2023-08-04 at 1 39 19 PM" src="https://github.com/growthbook/growthbook/assets/113377031/4b261fb6-d9a1-47b4-a613-2dc62dbd1822">

#### Add/Edit Feature modal

<img width="1049" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/a9f08ba3-43e0-427f-9557-5eb3bc7ca0f1">

#### Add/Edit Experiment modal

<img width="1039" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/55e288a4-0fb9-4e3d-8699-d422ff131655">

#### Datasources page

<img width="1209" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/29aac2b7-5892-42b5-8c98-e3d0ad161a77">



### Storybook Screenshots

When you visit the page and you do not have a demo project.

<img width="1299" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/6e0c657e-4ae3-48e6-96f6-254adf82c53d">

When you successfully create the demo project.

<img width="1287" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/b97252cb-701b-4536-990a-8c6dd3f0e7aa">

When you do a fresh load of the page and the demo project already exists.

<img width="1288" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/9f159c12-266b-4559-a8ec-b945eec6b5e2">

When you try to delete the already-existing demo project.

<img width="1295" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/fab564e0-2a00-441a-84a0-6b3a79c86ee2">


When you visit the page and it's loading still.

<img width="1289" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/ed63133f-1d24-4bb7-836e-8ab6123b250d">

lukesonnet commented on packages/front-end/components/Experiment/NewExperimentForm.tsx:
```diff
@@ -252,6 +255,13 @@ const NewExperimentForm: FC<NewExperimentFormProps> = ({
+
+        {currentProjectIsDemo && (
+          <div className="alert alert-warning">
+            You are creating an experiment under the demo datasource project.
```
Should we add "This experiment will be deleted when the demo datasource project is deleted."

To here and the other similar alerts?

tinahollygb replied:
Change made here 667afcdbf13a32e1f5d4c329a9cc7a8c3e7d5816

lukesonnet commented on packages/front-end/components/Metrics/MetricForm/index.tsx:
```diff
@@ -550,6 +563,11 @@ const MetricForm: FC<MetricFormProps> = ({
         >
+          {isExclusivelyForDemoDatasourceProject && (
+            <div className="alert alert-warning">
+              You are creating a metric under the demo datasource project.
```
One thing we can do to make this even safer is prevent them from using a different datasource in this project.

So I suggest 2 changes:
1. When creating the `form`, when in the demo datasource project, set the default to the demo datasource
2. In the Data Source selector, only let them select the demo datasource

tinahollygb replied:
I believe these changes mentioned here address these comments: https://github.com/growthbook/growthbook/pull/1499#issuecomment-1679612849 @lukesonnet

lukesonnet commented on packages/front-end/components/DemoDataSourcePage/DemoDataSourcePage.tsx:
```diff
@@ -0,0 +1,179 @@
+          All created resources will be deleted when the project is deleted.
+        </p>
+        <p>
+          If you accidentally delete something in the demo project and would
```
"accidentally delete one of our sample metrics, experiments or features and would like to restore it..."

tinahollygb replied:
Change made here 667afcdbf13a32e1f5d4c329a9cc7a8c3e7d5816

lukesonnet reviewed (commented):
Overall a few copy comments, but two main overall comments:

1.  On the DemoDataSourcePage, I think we need to make "what do I do next" after the page succeeds a lot clearer.

Maybe after success we can say "Show me the demo experiment" "Show me the demo feature" and have people click those buttons to go to the demo experiment or feature?

2. I agree we need a way to get back to this demo datasource page. I think in the demo page tooltip we can have a link to the demo datasource page.

tinahollygb commented:
@lukesonnet re:

> Maybe after success we can say "Show me the demo experiment" "Show me the demo feature" and have people click those buttons to go to the demo experiment or feature?

Let me know what those IDs are—are they hardcoded as well or do we need to look them up?

tinahollygb commented:
@lukesonnet re:

>  I think in the demo page tooltip we can have a link to the demo datasource page.

Yeah sounds good, I added a link here 1481dce1dd126fbf6dd7dd898b5e249aba415f96

<img width="245" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/7d8cb211-303f-4b64-99a9-f344edc3eba9">

lukesonnet commented:
> Let me know what those IDs are—are they hardcoded as well or do we need to look them up?

The feature ID is reliably `gbdemo-checkout-layout`, it had to be because of the way we link features and experiments.

The experiment ID will need to be looked up, but it should be the only one in this project.

tinahollygb commented:
The Add Metric modal has been updated when you're in the demo project:

- disables the project picker, defaulting to the demo project
- disables the datasource picker, defaulting to the demo datasource 

06a54a0952036e3d769d88b78e19e77ab94c34bc

<img width="1228" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/815bc219-d495-400a-af8d-694f93f93251">

tinahollygb commented:
@lukesonnet see fc68710283462f335e2e780c4e25097453191559 for a util abstraction I made to share the hardcoded feature ID please

tinahollygb commented:
Added links to the created resources on the demo datasource page (the buttons).

<img width="1244" alt="image" src="https://github.com/growthbook/growthbook/assets/113377031/55310450-e278-47e1-b4f7-26726183e62c">

tinahollygb commented:
Made changes in 120078f12bc9970ea5fdffe658370ced04aff359 to update the experiment to experiment-ref for a demo feature.

<img width="694" alt="Screen Shot 2023-08-17 at 12 23 31 PM" src="https://github.com/growthbook/growthbook/assets/113377031/3143b975-b4f2-4090-81f5-e1f7fef29520">

tinahollygb commented:
@lukesonnet @jdorn I believe I've addressed all feedback, please review.

jdorn commented on packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:
```diff
@@ -148,25 +147,23 @@ const DEMO_FEATURES: Omit<
-                value: "current",
-                weight: 0.3334,
+                value: "v0",
+                variationId: "current",
```
I think these are swapped.  `variationId` should be equal to the experiment `variations[i].id` field and `value` should be the string value of the feature flag for that variation.

tinahollygb replied:
Swapped them in this change f647b61c

jdorn commented on packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:
```diff
@@ -375,6 +357,28 @@ export const postDemoDatasourceProject = async (
+        // TODO: Ref: experiment-ref: Update this when updating demo data (features, experiments, etc.)
+        // Replaces the experimentId - Assumes the 2nd rule for all demo features is an ExperimentRefRule. If this is no longer true, we will need to change how the demo data is generated.
+        (featureToCreate.environmentSettings.production
+          .rules[1] as ExperimentRefRule).experimentId = experiments[i]?.id;
```
This seems very fragile.  Can we loop through the rules and check if they type is `experiment-ref` instead?  Also, look through every environment instead of hard-coding `production` here.

tinahollygb replied:
This was addressed [here](https://github.com/growthbook/growthbook/pull/1499#issuecomment-1688972938).

jdorn commented on packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:
```diff
@@ -148,25 +147,23 @@ const DEMO_FEATURES: Omit<
           },
           {
-            type: "experiment",
+            type: "experiment-ref",
```
This feature is hard-coding the environment as production.  What happens if the org doesn't have a `production` environment? (we create one by default when they create an org, but they can always delete or rename it)

Might be safer to look in their org for all of the defined environments and create a copy of these same rules for every environment.

tinahollygb replied:
Makes sense. I can build the rules programmatically based on the environments.

tinahollygb replied:
This was addressed [here](https://github.com/growthbook/growthbook/pull/1499#issuecomment-1688972938).

jdorn commented on packages/back-end/src/routers/demo-datasource-project/demo-datasource-project.controller.ts:
```diff
@@ -375,6 +357,28 @@ export const postDemoDatasourceProject = async (
+      DEMO_FEATURES.map(async (f, i) => {
+        const featureToCreate: FeatureInterface = {
+          ...f,
+          id: getDemoDataSourceFeatureId(),
```
If we're hardcoding the id, DEMO_FEATURES should not be an array.  Feature ids must be unique in an org, so adding more than one feature will break the code.

Same for DEMO_EXPERIMENTS.  We're hard-coding the trackingKey, which also should be unique per org.

lukesonnet replied:
Yeah. We can just change them not to be arrays for now (they were arrays originally because there were more than 1, but we can re-add that complexity if and when we need it).

tinahollygb replied:
So what's the requested change here? Remove the `DEMO_FEATURES` constant and inline the feature creation?

jdorn replied:
Yeah, that's probably easiest.  Or change it to `DEMO_FEATURE` and it's just a single feature, not an array.  But inlining is simpler.  We can refactor later if we decide to add more features

jdorn replied:
Same thing for `DEMO_EXPERIMENTS`, either inline or make it a single experiment, not an array

tinahollygb replied:
This was addressed [here](https://github.com/growthbook/growthbook/pull/1499#issuecomment-1688972938).

tinahollygb commented:
@jdorn in dc15b49c I've fixed the issues that were merged in PR #1487 but were somehow missed when you reviewed that PR but were caught here: 

- inlined the demo data
- removed the arrays for feature and experiment
- removed hardcoded production environment and instead iterate through the org environments
- iterate through the environments and rules

Let me know if any further changes are required.
